juju/juju
# Pass status recorder through api into backend for status history entries

In #17046 we added a new way to log the status history entries onto the files (like the debug-log) -- eventually to stop writing it into the db. Then we realized that the status update calls are getting picked up by the controller worker so the logs are written onto the controller model logs (e.g. `/var/log/juju/models/admin-controller-.....log`), which is not very useful because the status logs for entities need to be written per model for `show-status-log` to be accurate (imagine having two units with the same name on different models on the same controller).

This introduces a new function type `StatusHistoryRecorder` to keep the correct logger (in its environment in closure), and it is passed on the `StatusSetter` struct from the facade context (carrying the modelLogger for the entities) all the way through into the backend, so the correct recorder carrying the correct logger to be used to log the status history entry.

It looks like slightly more than a couple of source files got changed with this, but most of the changes are just adding extra parameter to the status setting calls (`SetStatus`, `SetInstanceStatus`, etc), along with tests and mocks, etc.

## Checklist




## QA steps

```
$ juju bootstrap deleteme
$ juju add-model tst
$ juju deploy ubuntu

After it deploys double check the statues are updated
$ juju show-status-log ubuntu/0
Time                        Type       Status       Message
28 Mar 2024 12:16:41-06:00  workload   waiting      waiting for machine
28 Mar 2024 12:16:41-06:00  juju-unit  allocating
28 Mar 2024 12:24:15-06:00  workload   waiting      installing agent
28 Mar 2024 12:24:17-06:00  workload   waiting      agent initialising
28 Mar 2024 12:28:56-06:00  workload   maintenance  installing charm software
28 Mar 2024 12:28:56-06:00  juju-unit  executing    running install hook
28 Mar 2024 12:29:10-06:00  juju-unit  executing    running leader-elected hook
28 Mar 2024 12:29:11-06:00  workload   active
28 Mar 2024 12:29:22-06:00  juju-unit  executing    running config-changed hook
28 Mar 2024 12:29:34-06:00  juju-unit  executing    running start hook
28 Mar 2024 12:29:38-06:00  juju-unit  idle
```

Keep in mind that these statuses are still coming from the DB, as the `show-status-log` command is not wired up to use the log files yet (actually it is, but not on this branch, in my other local branch, I just need to push it).

To see if this change is working, manually check to see if status logs are written in the model logs as follows:

```
 $ juju ssh -m controller 0

 $ cd /var/log/juju/models
 $ grep -rn "domain\":\"status"
```

You should see entries look like the following:
```
admin-tst-a9da8d.log:4:{"timestamp":"2024-03-28T18:16:04.029870521Z","entity":"machine-0","level":"INFO","module":"juju.status","location":"status.go:430","message":"","labels":{"domain":"status","id":"e","kind":"model","value":"available"}}
admin-tst-a9da8d.log:11:{"timestamp":"2024-03-28T18:16:50.126425839Z","entity":"machine-0","level":"INFO","module":"juju.status","location":"status.go:430","message":"starting","labels":{"domain":"status","id":"0","kind":"machine-instance","value":"allocating"}}
admin-tst-a9da8d.log:15:{"timestamp":"2024-03-28T18:16:54.406609275Z","entity":"machine-0","level":"INFO","module":"juju.status","location":"status.go:430","message":"acquiring LXD image","labels":{"domain":"status","id":"0","kind":"machine-instance","value":"allocating"}}
admin-tst-a9da8d.log:16:{"timestamp":"2024-03-28T18:16:54.5598391Z","entity":"machine-0","level":"INFO","module":"juju.status","location":"status.go:430","message":"Creating container","labels":{"domain":"status","id":"0","kind":"machine-instance","value":"allocating"}}
admin-tst-a9da8d.log:17:{"timestamp":"2024-03-28T18:16:55.44279424Z","entity":"machine-0","level":"INFO","module":"juju.status","location":"status.go:430","message":"starting","labels":{"domain":"status","id":"1","kind":"machine-instance","value":"allocating"}}
admin-tst-a9da8d.log:22:{"timestamp":"2024-03-28T18:16:56.046164186Z","entity":"machine-0","level":"INFO","module":"juju.status","location":"status.go:430","message":"acquiring LXD image","labels":{"domain":"status","id":"1","kind":"machine-instance","value":"allocating"}}
admin-tst-a9da8d.log:23:{"timestamp":"2024-03-28T18:16:57.831548182Z","entity":"machine-0","level":"INFO","module":"juju.status","location":"status.go:430","message":"Creating container","labels":{"domain":"status","id":"1","kind":"machine-instance","value":"allocating"}}
admin-tst-a9da8d.log:43:{"timestamp":"2024-03-28T18:24:15.662300159Z","entity":"machine-0","level":"INFO","module":"juju.status","location":"status.go:430","message":"installing agent","labels":{"domain":"status","id":"ubuntu/2","kind":"unit-workload","value":"waiting"}}
```

Note that the logs are in the `admin-tst-a9da8d.log` file, and not `admin-controller-a9da8d.log` file.

## Links

**Jira card:** JUJU-5789
